tests/common/plugins/pdu_controller/pdu_manager.py

```diff
@@ -0,0 +1,264 @@
+    and reference outlet buy outlet dictionary directly. With this,
+    we could enable different way to identify outlet, e.g. with the
+    outlet number from graph.
+
+    It also intended to create a smooth transition from defining
+    PDU in inventory to defining PDU in connection graph. Data in
+    graph is preferred, but if graph data is missing, existing
+    inventory data will be used.
+
+    PDU manager implements the same base PDU controller APIs and
+    collect status from and distribute operations to individual PDU
+    controllers.
+"""
+
+import logging
+import copy
+from snmp_pdu_controllers import get_pdu_controller
+
+logger = logging.getLogger(__name__)
+
+class PduManager():
+    def __init__(self, dut_hostname):
+        """
+            dut_hostname is the target DUT host name. The dut
+            defines which PDU(s) and outlet(s) it connected to.
+
+            It is NOT the PDU host name. PDU host name is defined
+            either in graph or in inventory and associated with
+            the DUT.
+        """
+        self.dut_hostname = dut_hostname
+        """
+            controlers is an array of controller dictionaries with
+            following information:
+            {
+                'psu_name'   : name of the PSU on DUT,
+                'host'       : controller_IP_address,
+                'controller' : controller instance,
+                'outlets'    : cached outlet status,
+                'psu_peer'   : psu peer information,
+            }
+        """
+        self.controllers = []
+
+
+    def _update_outlets(self, outlets, pdu_index):
+        for outlet in outlets:
+            outlet['pdu_index'] = pdu_index
+            outlet['pdu_name'] = self.controllers[pdu_index]['psu_peer']['peerdevice']
+
+
+    def add_controller(self, psu_name, psu_peer, pdu_vars):
+        """
+            Add a controller to be managed.
+            Sampel psu_peer:
+            {
+                "peerdevice": "pdu-107",
+                "HwSku": "Sentry",
+                "Protocol": "snmp",
+                "ManagementIp": "10.0.0.107",
+                "Type": "Pdu",
+                "peerport": "39"
+            }
+        """
+        if 'Protocol' not in psu_peer or 'ManagementIp' not in psu_peer:
+            logger.info('psu_peer {} missing critical inforamtion'.format(psu_peer))
+            return
+
+        if psu_peer['Protocol'] != 'snmp':
+            logger.warning('Controller protocol {} is not supported'.format(protocol))
+            return
+
+        controller = None
+        pdu_ip = psu_peer['ManagementIp']
+        shared_pdu = False
+        for pdu in self.controllers:
+            if psu_name in pdu:
+                logger.warning('PSU {} already has a pdu definition'.format(psu_name))
+                return
+            if pdu_ip == pdu['host']:
+                shared_pdu = True # Sharing controller with another outlet
```

Comment:
Per PEP8, there should be a minimum of 2 spaces before the `#` of a trailing comment. `autopep8` should also catch this.

Reply:
oh, didn't run against this file. But there are 2. Looks like on on the web page :-) And pep8 didn't complain.

Reply:
Strange. I only see one. :/

Reply:
Figured it out. I wasn't familiar with the tool. I thought the tool would complain, but it silently made some editing.

Reply:
That's due to the `--in-place` option. Without that option, it will just output the diff of the changes it would make.